File: QuickSort.py
def Partition(Array, left, right):
    key = Array[left]
    m = left
    for i in range(left + 1, right + 1):
        if Array[i] <= key:
            m += 1
            Array[i], Array[m] = Array[m], Array[i]
    Array[left], Array[m] = Array[m], Array[left]
    return m

def QuickSort(Array, left, right):
    if left < right:
        m = Partition(Array, left, right)
        QuickSort(Array, left, m - 1)
        QuickSort(Array, m + 1, right)

File: C_code/QuickSort.c
#include "QuickSort.h"
#include <stdio.h>

int partition(int *array, int left, int right)
{
	int key, m, i;
  
	key = array[left];	
	m   = left;
	
	for (i = left + 1; i < right + 1; i++) {
		if(array[i] <= key)
		{
			++m;
			swap(&array[i], &array[m]);
		}
	}
	
	swap(&array[left], &array[m]);
	return m;
}

void QuickSort(int *array, int left, int right)
{
	int m;
	
	if (left < right)
	{
		m = partition(array, left, right);
		QuickSort(array, left, m - 1);
		QuickSort(array, m + 1, right);
	}
}

void swap(int *a, int *b)
{
	int temp;
	
	temp = *a;
	*a   = *b;
	*b	 = temp;
}
